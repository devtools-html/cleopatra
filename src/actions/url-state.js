/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';

export function urlStateHasChanged(): Action {
  return {
    type: 'URL_STATE_HAS_CHANGED',
  };
}
