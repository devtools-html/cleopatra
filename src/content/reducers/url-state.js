// @flow
import { combineReducers } from 'redux';
import { defaultThreadOrder } from '../profile-data';
import { createSelector } from 'reselect';
import { urlFromState } from '../url-handling';
import * as RangeFilters from '../range-filters';

import type { ThreadIndex } from '../../common/types/profile';
import type { StartEndRange } from '../../common/types/units';
import type { Action, CallTreeFiltersPerThread, CallTreeFilter, DataSource } from '../actions/types';
import type { State, URLState, Reducer } from './types';

function dataSource(state: DataSource = 'none', action: Action) {
  switch (action.type) {
    case 'WAITING_FOR_PROFILE_FROM_FILE':
      return 'from-file';
    case 'PROFILE_PUBLISHED':
      return 'public';
    default:
      return state;
  }
}

function hash(state: string = '', action: Action) {
  switch (action.type) {
    case 'PROFILE_PUBLISHED':
      return action.hash;
    default:
      return state;
  }
}

function selectedTab(state: string = 'calltree', action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_TAB':
      return action.selectedTab;
    default:
      return state;
  }
}

function rangeFilters(state: StartEndRange[] = [], action: Action) {
  switch (action.type) {
    case 'ADD_RANGE_FILTER': {
      const { start, end } = action;
      return [...state, { start, end }];
    }
    case 'POP_RANGE_FILTERS':
      return state.slice(0, action.firstRemovedFilterIndex);
    default:
      return state;
  }
}

function selectedThread(state: ThreadIndex = 0, action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_THREAD':
      return action.selectedThread;
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_FILE': {
      // When loading in a brand new profile, select either the GeckoMain [tab] thread,
      // or the first thread in the thread order. For profiles from the Web, the
      // selectedThread has already been initialized from the URL and does not require
      // looking at the profile.
      const contentThreadId = action.profile.threads.findIndex(thread => {
        return thread.name === 'GeckoMain' && thread.processType === 'tab';
      });
      return contentThreadId !== -1 ? contentThreadId : defaultThreadOrder(action.profile.threads)[0];
    }
    default:
      return state;
  }
}

function callTreeSearchString(state: string = '', action: Action) {
  switch (action.type) {
    case 'CHANGE_CALL_TREE_SEARCH_STRING':
      return action.searchString;
    default:
      return state;
  }
}

function callTreeFilters(state: CallTreeFiltersPerThread = {}, action: Action) {
  switch (action.type) {
    case 'ADD_CALL_TREE_FILTER': {
      const { threadIndex, filter } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: [...oldFilters, filter],
      });
    }
    case 'POP_CALL_TREE_FILTERS': {
      const { threadIndex, firstRemovedFilterIndex } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: oldFilters.slice(0, firstRemovedFilterIndex),
      });
    }
    default:
      return state;
  }
}

function jsOnly(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_JS_ONLY':
      return action.jsOnly;
    default:
      return state;
  }
}

function invertCallstack(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
      return action.invertCallstack;
    default:
      return state;
  }
}

function hidePlatformDetails(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_HIDE_PLATFORM_DETAILS':
      return action.hidePlatformDetails;
    default:
      return state;
  }
}

const urlStateReducer: Reducer<URLState> = (regularUrlStateReducer => (state: URLState, action: Action): URLState => {
  switch (action.type) {
    case '@@urlenhancer/updateURLState':
      return action.urlState;
    default:
      return regularUrlStateReducer(state, action);
  }
})(combineReducers({
  dataSource, hash, selectedTab, rangeFilters, selectedThread,
  callTreeSearchString, callTreeFilters, jsOnly, invertCallstack,
  hidePlatformDetails,
}));
export default urlStateReducer;

const getURLState = (state: State): URLState => state.urlState;

export const getDataSource = (state: State) => getURLState(state).dataSource;
export const getHash = (state: State) => getURLState(state).hash;
export const getRangeFilters = (state: State) => getURLState(state).rangeFilters;
export const getJSOnly = (state: State) => getURLState(state).jsOnly;
export const getHidePlatformDetails = (state: State) => getURLState(state).hidePlatformDetails;
export const getInvertCallstack = (state: State) => getURLState(state).invertCallstack;
export const getSearchString = (state: State) => getURLState(state).callTreeSearchString;
export const getSelectedTab = (state: State) => getURLState(state).selectedTab;
export const getSelectedThreadIndex = (state: State) => getURLState(state).selectedThread;
export const getCallTreeFilters = (state: State, threadIndex: ThreadIndex): CallTreeFilter[] => {
  return getURLState(state).callTreeFilters[threadIndex] || [];
};

export const getURLPredictor = createSelector(
  getURLState,
  (oldURLState: URLState) => actionOrActionList => {
    const actionList = ('type' in actionOrActionList) ? [actionOrActionList] : actionOrActionList;
    const newURLState = actionList.reduce(urlStateReducer, oldURLState);
    return urlFromState(newURLState);
  }
);

export const getRangeFilterLabels = createSelector(
  getRangeFilters,
  RangeFilters.getRangeFilterLabels
);
