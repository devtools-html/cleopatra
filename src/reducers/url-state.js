/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';
import escapeStringRegexp from 'escape-string-regexp';
import { createSelector } from 'reselect';
import { ensureExists } from '../utils/flow';
import { urlFromState } from '../app-logic/url-handling';
import * as CommittedRanges from '../profile-logic/committed-ranges';

import type { ThreadIndex, Pid } from '../types/profile';
import type { TrackIndex } from '../types/profile-derived';
import type { StartEndRange } from '../types/units';
import type {
  TransformStacksPerThread,
  TransformStack,
} from '../types/transforms';
import type {
  Action,
  DataSource,
  ImplementationFilter,
  TimelineType,
} from '../types/actions';
import type { State, UrlState, Reducer } from '../types/reducers';
import type { TabSlug } from '../app-logic/tabs-handling';

// Pre-allocate an array to help with strict equality tests in the selectors.
const EMPTY_TRANSFORM_STACK = [];

function dataSource(state: DataSource = 'none', action: Action) {
  switch (action.type) {
    case 'WAITING_FOR_PROFILE_FROM_FILE':
      return 'from-file';
    case 'PROFILE_PUBLISHED':
      return 'public';
    default:
      return state;
  }
}

function hash(state: string = '', action: Action) {
  switch (action.type) {
    case 'PROFILE_PUBLISHED':
      return action.hash;
    default:
      return state;
  }
}

function profileUrl(state: string = '', action: Action) {
  switch (action.type) {
    default:
      return state;
  }
}

function selectedTab(state: TabSlug = 'calltree', action: Action): TabSlug {
  switch (action.type) {
    case 'CHANGE_SELECTED_TAB':
    case 'SELECT_TRACK':
      return action.selectedTab;
    default:
      return state;
  }
}

function committedRanges(state: StartEndRange[] = [], action: Action) {
  switch (action.type) {
    case 'COMMIT_RANGE': {
      const { start, end } = action;
      return [...state, { start, end }];
    }
    case 'POP_COMMITTED_RANGES':
      return state.slice(0, action.firstPoppedFilterIndex);
    default:
      return state;
  }
}

function selectedThread(
  state: ThreadIndex | null = null,
  action: Action
): ThreadIndex | null {
  switch (action.type) {
    case 'CHANGE_SELECTED_THREAD':
    case 'SELECT_TRACK':
    case 'VIEW_PROFILE':
    case 'ISOLATE_PROCESS':
    case 'ISOLATE_PROCESS_MAIN_THREAD':
    case 'HIDE_GLOBAL_TRACK':
    case 'HIDE_LOCAL_TRACK':
    case 'ISOLATE_LOCAL_TRACK':
      // Only switch to non-null selected threads.
      return (action.selectedThreadIndex: ThreadIndex);
    default:
      return state;
  }
}

function callTreeSearchString(state: string = '', action: Action) {
  switch (action.type) {
    case 'CHANGE_CALL_TREE_SEARCH_STRING':
      return action.searchString;
    default:
      return state;
  }
}

function markersSearchString(state: string = '', action: Action) {
  switch (action.type) {
    case 'CHANGE_MARKER_SEARCH_STRING':
      return action.searchString;
    default:
      return state;
  }
}

function transforms(state: TransformStacksPerThread = {}, action: Action) {
  switch (action.type) {
    case 'ADD_TRANSFORM_TO_STACK': {
      const { threadIndex, transform } = action;
      const transforms = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: [...transforms, transform],
      });
    }
    case 'POP_TRANSFORMS_FROM_STACK': {
      const { threadIndex, firstPoppedFilterIndex } = action;
      const transforms = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: transforms.slice(0, firstPoppedFilterIndex),
      });
    }
    default:
      return state;
  }
}

function timelineType(
  state: TimelineType = 'category',
  action: Action
): TimelineType {
  switch (action.type) {
    case 'CHANGE_TIMELINE_TYPE':
      return action.timelineType;
    default:
      return state;
  }
}

/**
 * Represents the current filter applied to the stack frames, where it will show
 * frames only by implementation.
 */
function implementation(
  state: ImplementationFilter = 'combined',
  action: Action
) {
  switch (action.type) {
    case 'CHANGE_IMPLEMENTATION_FILTER':
      return action.implementation;
    default:
      return state;
  }
}

function invertCallstack(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
      return action.invertCallstack;
    default:
      return state;
  }
}

function globalTrackOrder(state: TrackIndex[] = [], action: Action) {
  switch (action.type) {
    case 'VIEW_PROFILE':
    case 'CHANGE_GLOBAL_TRACK_ORDER':
      return action.globalTrackOrder;
    default:
      return state;
  }
}

function hiddenGlobalTracks(
  state: Set<TrackIndex> = new Set(),
  action: Action
) {
  switch (action.type) {
    case 'VIEW_PROFILE':
    case 'ISOLATE_LOCAL_TRACK':
    case 'ISOLATE_PROCESS':
    case 'ISOLATE_PROCESS_MAIN_THREAD':
      return action.hiddenGlobalTracks;
    case 'HIDE_GLOBAL_TRACK': {
      const hiddenGlobalTracks = new Set(state);
      hiddenGlobalTracks.add(action.trackIndex);
      return hiddenGlobalTracks;
    }
    case 'SHOW_GLOBAL_TRACK': {
      const hiddenGlobalTracks = new Set(state);
      hiddenGlobalTracks.delete(action.trackIndex);
      return hiddenGlobalTracks;
    }
    default:
      return state;
  }
}

function hiddenLocalTracksByPid(
  state: Map<Pid, Set<TrackIndex>> = new Map(),
  action: Action
) {
  switch (action.type) {
    case 'VIEW_PROFILE':
      return action.hiddenLocalTracksByPid;
    case 'HIDE_LOCAL_TRACK': {
      const hiddenLocalTracksByPid = new Map(state);
      const hiddenLocalTracks = new Set(hiddenLocalTracksByPid.get(action.pid));
      hiddenLocalTracks.add(action.trackIndex);
      hiddenLocalTracksByPid.set(action.pid, hiddenLocalTracks);
      return hiddenLocalTracksByPid;
    }
    case 'SHOW_LOCAL_TRACK': {
      const hiddenLocalTracksByPid = new Map(state);
      const hiddenLocalTracks = new Set(hiddenLocalTracksByPid.get(action.pid));
      hiddenLocalTracks.delete(action.trackIndex);
      hiddenLocalTracksByPid.set(action.pid, hiddenLocalTracks);
      return hiddenLocalTracksByPid;
    }
    case 'ISOLATE_PROCESS_MAIN_THREAD':
    case 'ISOLATE_LOCAL_TRACK': {
      const hiddenLocalTracksByPid = new Map(state);
      hiddenLocalTracksByPid.set(action.pid, action.hiddenLocalTracks);
      return hiddenLocalTracksByPid;
    }
    default:
      return state;
  }
}

function localTrackOrderByPid(
  state: Map<Pid, TrackIndex[]> = new Map(),
  action: Action
) {
  switch (action.type) {
    case 'VIEW_PROFILE':
      return action.localTrackOrderByPid;
    case 'CHANGE_LOCAL_TRACK_ORDER': {
      const localTrackOrderByPid = new Map(state);
      localTrackOrderByPid.set(action.pid, action.localTrackOrder);
      return localTrackOrderByPid;
    }
    default:
      return state;
  }
}

function pathInZipFile(
  state: string | null = null,
  action: Action
): string | null {
  switch (action.type) {
    // Update the URL the moment the zip file is starting to be
    // processed, not when it is viewed. The processing is async.
    case 'PROCESS_PROFILE_FROM_ZIP_FILE':
      return action.pathInZipFile ? action.pathInZipFile : null;
    case 'RETURN_TO_ZIP_FILE_LIST':
      return null;
    default:
      return state;
  }
}

/**
 * These values are specific to an individual profile.
 */
const profileSpecific = combineReducers({
  selectedThread,
  globalTrackOrder,
  hiddenGlobalTracks,
  hiddenLocalTracksByPid,
  localTrackOrderByPid,
  implementation,
  invertCallstack,
  committedRanges,
  callTreeSearchString,
  markersSearchString,
  transforms,
  timelineType,
  // The timeline tracks used to be hidden and sorted by thread indexes, rather than
  // track indexes. The only way to migrate this information to tracks-based data is to
  // first retrieve the profile, so they can't be upgraded by the normal url upgrading
  // process. These value are only set by the locationToState function.
  legacyThreadOrder: (state: ThreadIndex[] | null = null) => state,
  legacyHiddenThreads: (state: ThreadIndex[] | null = null) => state,
});

/**
 * Provide a mechanism to wrap the UrlState reducer in a special function that can swap
 * out the entire UrlState with a new one coming from the History API. Also provide a
 * way to invalidate sections of the state based off of looking at multiple profiles.
 */
const wrapReducerInResetter = (
  regularUrlStateReducer: Reducer<UrlState>
): Reducer<UrlState> => {
  return (state, action) => {
    switch (action.type) {
      case 'UPDATE_URL_STATE':
        // A new URL came in because of a browser action, discard the current UrlState
        // and use the new one, which was probably serialized from the URL, or stored
        // in the history API.
        return action.newUrlState
          ? action.newUrlState
          : regularUrlStateReducer(undefined, action);
      case 'RETURN_TO_ZIP_FILE_LIST':
        // Invalidate all information that would be specific to an individual profile.
        return {
          ...regularUrlStateReducer(state, action),
          profileSpecific: profileSpecific(undefined, state),
        };
      default:
        return regularUrlStateReducer(state, action);
    }
  };
};

const urlStateReducer = wrapReducerInResetter(
  combineReducers({
    dataSource,
    hash,
    profileUrl,
    selectedTab,
    pathInZipFile,
    profileSpecific,
  })
);

export default urlStateReducer;

export const getUrlState = (state: State): UrlState => state.urlState;
export const getProfileSpecificState = (state: State) =>
  getUrlState(state).profileSpecific;

export const getDataSource = (state: State) => getUrlState(state).dataSource;
export const getHash = (state: State) => getUrlState(state).hash;
export const getProfileUrl = (state: State) => getUrlState(state).profileUrl;
export const getAllCommittedRanges = (state: State) =>
  getProfileSpecificState(state).committedRanges;
export const getImplementationFilter = (state: State) =>
  getProfileSpecificState(state).implementation;
export const getInvertCallstack = (state: State) =>
  getProfileSpecificState(state).invertCallstack;
export const getCurrentSearchString = (state: State) =>
  getProfileSpecificState(state).callTreeSearchString;
export const getSearchStrings = createSelector(
  getCurrentSearchString,
  searchString => {
    if (!searchString) {
      return null;
    }
    const result = searchString
      .split(',')
      .map(part => part.trim())
      .filter(part => part);

    if (result.length) {
      return result;
    }

    return null;
  }
);
export const getSearchStringsAsRegExp = createSelector(
  getSearchStrings,
  strings => {
    if (!strings || !strings.length) {
      return null;
    }
    const regexpStr = strings.map(escapeStringRegexp).join('|');
    return new RegExp(regexpStr, 'gi');
  }
);
export const getMarkersSearchString = (state: State) =>
  getProfileSpecificState(state).markersSearchString;
export const getSelectedTab = (state: State) => getUrlState(state).selectedTab;
export const getSelectedThreadIndexOrNull = (state: State) =>
  getProfileSpecificState(state).selectedThread;
export const getSelectedThreadIndex = (state: State) => {
  const threadIndex = getSelectedThreadIndexOrNull(state);
  if (threadIndex === null) {
    throw new Error(
      'Attempted to get a thread index before a profile was loaded.'
    );
  }
  return threadIndex;
};
export const getTransformStack = (
  state: State,
  threadIndex: ThreadIndex
): TransformStack => {
  return (
    getProfileSpecificState(state).transforms[threadIndex] ||
    EMPTY_TRANSFORM_STACK
  );
};

export const getTimelineType = (state: State): TimelineType =>
  getProfileSpecificState(state).timelineType;

export const getLegacyThreadOrder = (state: State) =>
  getProfileSpecificState(state).legacyThreadOrder;
export const getLegacyHiddenThreads = (state: State) =>
  getProfileSpecificState(state).legacyHiddenThreads;
export const getGlobalTrackOrder = (state: State) =>
  getProfileSpecificState(state).globalTrackOrder;
export const getHiddenGlobalTracks = (state: State) =>
  getProfileSpecificState(state).hiddenGlobalTracks;
export const getHiddenLocalTracksByPid = (state: State) =>
  getProfileSpecificState(state).hiddenLocalTracksByPid;
export const getHiddenLocalTracks = (state: State, pid: Pid) =>
  ensureExists(
    getHiddenLocalTracksByPid(state).get(pid),
    'Unable to get the hidden tracks from the given pid'
  );
export const getLocalTrackOrderByPid = (state: State) =>
  getProfileSpecificState(state).localTrackOrderByPid;
export const getLocalTrackOrder = (state: State, pid: Pid) =>
  ensureExists(
    getLocalTrackOrderByPid(state).get(pid),
    'Unable to get the track order from the given pid'
  );

export const getUrlPredictor = createSelector(
  getUrlState,
  (oldUrlState: UrlState) => (actionOrActionList: Action | Action[]) => {
    const actionList: Action[] = Array.isArray(actionOrActionList)
      ? actionOrActionList
      : [actionOrActionList];
    const newUrlState = actionList.reduce(urlStateReducer, oldUrlState);
    return urlFromState(newUrlState);
  }
);

export const getPathInZipFileFromUrl = (state: State) =>
  getUrlState(state).pathInZipFile;

/**
 * For now only provide a name for a profile if it came from a zip file.
 */
export const getProfileName: State => null | string = createSelector(
  getPathInZipFileFromUrl,
  pathInZipFile => {
    if (!pathInZipFile) {
      return null;
    }
    const pathParts = pathInZipFile.split('/');
    return pathParts[pathParts.length - 1];
  }
);

export const getCommittedRangeLabels = createSelector(
  getAllCommittedRanges,
  CommittedRanges.getCommittedRangeLabels
);
