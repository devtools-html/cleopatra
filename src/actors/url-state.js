/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { getUrlState } from '../reducers/url-state';
import {
  urlStateHasChanged,
  updateUrlState,
  show404,
} from '../actions/url-state';
import { urlFromState, stateFromLocation } from '../url-handling';

import type { Dispatch } from '../types/store';
import type { State, UrlState } from '../types/reducers';

let previousUrlState = null;
export default function stateWatcher(state: State, dispatch: Dispatch) {
  function updateState() {
    if (window.history.state) {
      dispatch(updateUrlState(window.history.state));
    } else {
      try {
        const urlState = stateFromLocation(window.location);
        // replacing the existing location ensures we get the latest format for the URL
        updateLocation(urlState, true);
        dispatch(updateUrlState(urlState));
      } catch (e) {
        console.error(e);
        show404(window.location.pathname + window.location.search);
      }
    }
  }

  function initialSetup() {
    updateState();
    window.addEventListener('popstate', updateState);
  }

  function updateLocation(urlState: UrlState, replace: boolean = false) {
    const newUrl = urlFromState(urlState);
    const operation = replace
      ? window.history.replaceState.bind(window.history)
      : window.history.pushState.bind(window.history);
    if (newUrl !== window.location.pathname + window.location.search) {
      operation(urlState, document.title, newUrl);
    }
  }

  const newUrlState = getUrlState(state);
  if (!previousUrlState) {
    // first run
    previousUrlState = newUrlState;
    initialSetup();
  } else if (newUrlState !== previousUrlState) {
    previousUrlState = newUrlState;
    updateLocation(newUrlState);
    dispatch(urlStateHasChanged());
  }
}
