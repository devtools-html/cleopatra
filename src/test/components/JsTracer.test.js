/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import {
  TIMELINE_MARGIN_LEFT,
  TIMELINE_MARGIN_RIGHT,
} from '../../app-logic/constants';
import JsTracer from '../../components/js-tracer';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import mockRaf from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  getBoundingBox,
  addRootOverlayElement,
  removeRootOverlayElement,
} from '../fixtures/utils';
import { getProfileWithJsTracerEvents } from '../fixtures/profiles/make-profile';
import { getShowJsTracerSummary } from '../../selectors/url-state';
jest.useFakeTimers();

const GRAPH_BASE_WIDTH = 200;
const GRAPH_WIDTH =
  GRAPH_BASE_WIDTH + TIMELINE_MARGIN_LEFT + TIMELINE_MARGIN_RIGHT;
const GRAPH_HEIGHT = 300;

describe('StackChart', function() {
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  function setup({
    skipLoadingScreen,
    events,
  }: {
    skipLoadingScreen: boolean,
    events: Array<*>,
  }) {
    const flushRafCalls = mockRaf();
    const ctx = mockCanvasContext();

    jest
      .spyOn(HTMLCanvasElement.prototype, 'getContext')
      .mockImplementation(() => ctx);

    jest
      .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
      .mockImplementation(() => getBoundingBox(GRAPH_WIDTH, GRAPH_HEIGHT));

    const profile = getProfileWithJsTracerEvents(events);

    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;

    const view = mount(
      <Provider store={store}>
        <JsTracer />
      </Provider>
    );
    if (skipLoadingScreen) {
      // Have React show the loading screen.
      flushRafCalls();
      // Have React show our new component, and schedule a draw call.
      flushRafCalls();
      // Now flush once more to actually draw to the screen.
      flushRafCalls();
      view.update();
    }

    const getJsTracerChartCanvas = () => view.find('canvas').first();
    const getChangeJsTracerSummaryCheckbox = () =>
      view.find('.jsTracerSettingsCheckbox');

    return {
      dispatch,
      getState,
      view,
      ctx,
      flushRafCalls,
      getJsTracerChartCanvas,
      getChangeJsTracerSummaryCheckbox,
    };
  }

  const simpleTracerEvents = [
    ['https://mozilla.org', 0, 20],
    ['Interpreter', 1, 19],
    ['IonMonkey', 2, 18],
  ];

  it('only computes the chart after the first tick', () => {
    const { view, flushRafCalls } = setup({
      skipLoadingScreen: false,
      events: simpleTracerEvents,
    });
    expect(view.find('canvas').length).toEqual(0);
    // Flush twice, as the canvas defers until after React updates.
    flushRafCalls();
    flushRafCalls();
    view.update();
    expect(view.find('canvas').length).toEqual(1);
  });

  it('matches the snapshot for the loading screen', () => {
    const { view } = setup({
      skipLoadingScreen: false,
      events: simpleTracerEvents,
    });
    expect(view).toMatchSnapshot();
  });

  it('matches the snapshot for empty reasons screen', () => {
    const { view } = setup({
      skipLoadingScreen: true,
      events: [],
    });
    expect(view).toMatchSnapshot();
  });

  it('matches the snapshot for the chart', () => {
    const { view } = setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    expect(view).toMatchSnapshot();
  });

  it('matches the snapshot a simple chart render', () => {
    const { ctx } = setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    expect(ctx.__flushDrawLog()).toMatchSnapshot();
  });

  it('can change to a summary view', function() {
    const { getChangeJsTracerSummaryCheckbox, getState } = setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    expect(getShowJsTracerSummary(getState())).toEqual(false);
    getChangeJsTracerSummaryCheckbox().simulate('change');
    expect(getShowJsTracerSummary(getState())).toEqual(true);
  });

  it('matches the snapshot for an inverted draw call', function() {
    const { getChangeJsTracerSummaryCheckbox, ctx } = setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    getChangeJsTracerSummaryCheckbox().simulate('change');
    expect(ctx.__flushDrawLog()).toMatchSnapshot();
  });
});
