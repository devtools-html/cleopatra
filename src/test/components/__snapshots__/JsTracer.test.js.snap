// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackChart matches the snapshot a simple chart render 1`] = `
Array [
  Array [
    "set font",
    "10px sans-serif",
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "…",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    150,
    1,
    200,
    13,
  ],
  Array [
    "measureText",
    "https://mozilla.org",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "https://mozilla.org",
    153,
    11,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    160,
    17,
    179,
    13,
  ],
  Array [
    "measureText",
    "Interpreter",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "Interpreter",
    163,
    27,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    339,
    18,
    0.9999999999999745,
    11,
  ],
  Array [
    "fillRect",
    170,
    33,
    160,
    13,
  ],
  Array [
    "measureText",
    "IonMonkey",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "IonMonkey",
    173,
    43,
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "measureText",
    "Tracing Information",
  ],
  Array [
    "fillText",
    "Tracing Information",
    5,
    11,
  ],
]
`;

exports[`StackChart matches the snapshot for an inverted draw call 1`] = `
Array [
  Array [
    "set font",
    "10px sans-serif",
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "…",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    150,
    1,
    200,
    13,
  ],
  Array [
    "measureText",
    "https://mozilla.org",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "https://mozilla.org",
    153,
    11,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    160,
    17,
    179,
    13,
  ],
  Array [
    "measureText",
    "Interpreter",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "Interpreter",
    163,
    27,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    339,
    18,
    0.9999999999999745,
    11,
  ],
  Array [
    "fillRect",
    170,
    33,
    160,
    13,
  ],
  Array [
    "measureText",
    "IonMonkey",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "IonMonkey",
    173,
    43,
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "measureText",
    "Tracing Information",
  ],
  Array [
    "fillText",
    "Tracing Information",
    5,
    11,
  ],
]
`;

exports[`StackChart matches the snapshot for empty reasons screen 1`] = `
<div
  className="jsTracer"
>
  <div
    className="EmptyReasons"
  >
    <h3>
      The 
      js-tracer
       is empty for “
      Empty
      ”
    </h3>
    <p>
      This thread contains no JS tracer information.
    </p>
  </div>
</div>
`;

exports[`StackChart matches the snapshot for the chart 1`] = `
<div
  className="jsTracer"
>
  <div
    className="jsTracerSettings"
  >
    <ul
      className="jsTracerSettingsList"
    >
      <li
        className="jsTracerSettingsListItem"
      >
        <label
          className="jsTracerSettingsLabel"
        >
          <input
            checked={false}
            className="jsTracerSettingsCheckbox"
            onChange={[Function]}
            type="checkbox"
          />
           Show only self time
        </label>
      </li>
    </ul>
  </div>
  <div
    className="chartViewport"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onWheel={[Function]}
    tabIndex={0}
  >
    <div>
      <canvas
        className="chartCanvas jsTracerCanvasFadeIn jsTracerCanvas jsTracerCanvasDrawn"
        onDoubleClick={[Function]}
        onMouseDown={[Function]}
        onMouseMove={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      />
    </div>
    <div
      className="chartViewportShiftScroll hidden"
    >
      Zoom Chart:
      <kbd
        className="chartViewportShiftScrollKbd"
      >
        Shift
      </kbd>
      <kbd
        className="chartViewportShiftScrollKbd"
      >
        Scroll
      </kbd>
    </div>
  </div>
</div>
`;

exports[`StackChart matches the snapshot for the loading screen 1`] = `
<div
  className="jsTracer"
>
  <div
    className="jsTracerSettings"
  >
    <ul
      className="jsTracerSettingsList"
    >
      <li
        className="jsTracerSettingsListItem"
      >
        <label
          className="jsTracerSettingsLabel"
        >
          <input
            checked={false}
            className="jsTracerSettingsCheckbox"
            onChange={[Function]}
            type="checkbox"
          />
           Show only self time
        </label>
      </li>
    </ul>
  </div>
  <div
    className="jsTracerLoader"
  >
    Re-constructing tracing information from
     
    3
     events. This might take a moment.
  </div>
</div>
`;
